component/discovery/discovery.go

```diff
@@ -0,0 +1,67 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/agent/component/metrics/scrape"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// RunDiscovery is a utility for consuming and forwarding target groups from a discoverer.
+// It will handle collating targets (and clearing), as well as time based throttling of updates.
+// The channel provided should be the same channel passed to the Discovery's Run method.
+func RunDiscovery(ctx context.Context, ch <-chan []*targetgroup.Group, f func([]scrape.Target)) {
```

Comment:
It feels like this should be `RunDiscovery(ctx context.Context, d discovery.Discoverer, f func([]scrape.Target))`.

Reply:
The problem with that is lifecycle of the Discoverer. If the component updates, we will make a new Discoverer. 

But we don't need to make a new channel. This loop can continue uninterrupted.

But maybe restarting this is viable too. Lemme see if that works.

Reply:
On second thought, I think we **need** to restart this loop on update. Otherwise you could get orphaned cached targets.